src/shims/foreign_items/posix.rs

```diff
@@ -248,9 +275,10 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 this.write_null(dest)?;
             }
             "pthread_atfork" => {
+                check_arg_count(args, 3)?;
                 let _prepare = this.read_scalar(args[0])?.not_undef()?;
                 let _parent = this.read_scalar(args[1])?.not_undef()?;
-                let _child = this.read_scalar(args[1])?.not_undef()?;
+                let _child = this.read_scalar(args[2])?.not_undef()?;
```

Comment:
This seems to have just been a latent bug.

Reply:
Indeed that was probably a typo on my side, good catch.
Now I wonder if we could prevent this through some type-based magic... like `let [prepare, parent, child] = check_arg_count::<3>(args)` or so. In fact, I think libstd has `TryFrom` instances to turn slices into arrays, so maybe we could use that? Something [like this snippet](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=079d641eabe4a45517654c586f2d7773). Maybe with a macro to make it less repetetive, though that may not be worth it.

We can leave that to a future PR though if you want.

Reply:
That was my original thought reading this issue, and I'd probably prefer that version ultimately. At least one version I've tried managed to ICE, so I don't know if using const generics for this is worth it just yet?

```
// in helpers.rs under
// pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
    fn read_scalars<const N: usize>(&mut self, args: &[OpTy<'tcx, Tag>]) -> InterpResult<'tcx, [Scalar<Tag>; N]> {
        let this = self.eval_context_mut();
        if args.len() != N {
            throw_ub_format!("incorrect number of arguments, got {}, needed {}", args.len(), N);
        }
        let mut scalars = [Scalar::<Tag>::from_u8(0); N];
        for (i, &arg) in args.iter().enumerate() {
            scalars[i] = this.read_scalar(arg)?.not_undef()?;
        }
        Ok(scalars)
    }

// then
--- a/src/shims/intrinsics.rs
+++ b/src/shims/intrinsics.rs
@@ -416,8 +416,10 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx

             "offset" => {
                 check_arg_count(args, 2)?;
-                let offset = this.read_scalar(args[1])?.to_machine_isize(this)?;
-                let ptr = this.read_scalar(args[0])?.not_undef()?;
+                let [offset, ptr] = this.read_scalars::<2usize>(args);
+                let offset = offset.to_machine_isize(this)?;
                 let result_ptr = this.pointer_offset_inbounds(ptr, substs.type_at(0), offset)?;
                 this.write_scalar(result_ptr, dest)?;
             }
```

Reply:
Yeah, const-generic is pretty ICE-prone... but if we restrict ourselves to the patterns libstd uses, it should not be too bad I hope. We could always fall back to a macro here.

However, please avoid this pattern:
```rust
        let mut scalars = [Scalar::<Tag>::from_u8(0); N];
        for (i, &arg) in args.iter().enumerate() {
            scalars[i] = this.read_scalar(arg)?.not_undef()?;
        }
```
I'd rather not risk accidentally returning the default value. This is exactly what we have [`TryFrom` instances for](https://doc.rust-lang.org/nightly/src/core/array/mod.rs.html#137-147).

Reply:
> However, please avoid this pattern:

I was indeed using TryFrom in other versions of this, but those versions wouldn't compile either. This is merely the version that ICE'd.

Reply:
I think the ICE is from passing `2usize` to a method's const generic parameter (we've been meaning to ban that sort of thing, it's unsupportable).

Can you try to rely on inference instead? It should "just work".

Reply:
@eddyb Tried `this.read_scalars::<2>(args);` and `this.read_scalars(args);`, they did not seem to compile. But I could try again. One question I have is if one of these versions using const generics does manage to work can we just depend on it continuing to work? Or might it start breaking in the future?

Reply:
Update: I've got a version that works, redoing the PR with that.